mapping_parenting_tech/analysis/prototyping/play_store/fn_tester.py

```diff
@@ -0,0 +1,140 @@
+len(appIds)
+
+# %% [markdown]
+# ### Saving app ids and loading saved app ids
+# App ids are saved to a single CSV file `outputs/data/all_app_ids.csv` with one app id per line
+
+# %%
+# test saving new apps - dry run allows us to do this without updating the file itself
+# function returns complete list of new apps (i.e., not just those that have been added)
+new_appIds = psu.update_all_app_id_list(appIds, dry_run=False)
+new_appIds
+
+# %%
+# load all app ids
+all_appIds = psu.load_all_app_ids()
+len(all_appIds)
+
+# %%
+# check if an app is in the list of all app ids - here we're using the results of the snowball above
+compare_list = psu.is_app_in_list(appIds)
+
+# enumerate through compare_list to identify apps that aren't present
+print(f"{compare_list.count(False)} apps are not already saved")
+for i, present in enumerate(compare_list):
+    if not present:
+        print(appIds[i])
+
+# %% [markdown]
+# ## Working with app details
+# App details can be downloaded into memory in a standalone fashion or saved to and loaded from disk via the JSON file: `outputs/data/all_app_details.json`. This file is updated/synchronised according to the saved app ids.
+
+# %%
+# get the app details for the list of apps in `app_list`
+
+app_list = [
+    "com.amboss.medical.knowledge",
+    "com.excelatlife.depression",
+    "au.com.sightwords.parrotfish.lite",
+    "com.kitefaster.whitenoise",
+    "com.apololo.brightest.flashlight",
+]
+app_list_details = psu.get_playstore_app_details(app_list)
+app_list_details["com.amboss.medical.knowledge"]["title"]
+
+# %%
+# load all app details into a dict
+all_app_details = psu.load_all_app_details()
+len(all_app_details)
+
+# %%
+# update the details for apps whose ids are already saved - this returns the details for all of the apps in question
+all_app_details = psu.update_all_app_details()
+len(all_app_details)
+
+# %% [markdown]
+# ## Working with app reviews
+# App reviews are saved into `outputs/data/app_reviews` but are loaded into a consolidated Pandas DataFrame. Due to the potential number of reviews for each app, downloading can take some time
+
+# %%
+# get some app reviews for a given app - if there are more than 200 reviews, this will return the most recent
+# 200 reviews. Note that nothing is saved to disc in this procedure
+my_app_reviews, _ = psu.get_playstore_app_reviews("com.amboss.medical.knowledge")
+len(my_app_reviews)
+
+# %%
+# download reviews for apps in our example app list, forcing all reviews to be downloaded
+app_list = [
+    "com.amboss.medical.knowledge",
+    "com.excelatlife.depression",
+    "au.com.sightwords.parrotfish.lite",
+    "com.kitefaster.whitenoise",
+    "com.apololo.brightest.flashlight",
+]
+psu.save_playstore_app_list_reviews(app_list, force_download=True, run_quietly=False)
+
+# %%
+# get a list of app ids for those apps whose reviews haven't been downloaded
+apps_to_get = psu.list_missing_app_reviews()
```

Comment:
Is that expected that there was an error with loading one of the apps? (see below)

```
2022-02-25 17:41:10,591 - root - INFO - Loading files
1153it [00:21, 69.66it/s]
2022-02-25 17:41:32,317 - root - WARNING - Error loading /Users/karliskanders/Documents/mapping_parenting_tech/outputs/data/app_reviews/com.chineseskill.csv
3911it [01:15, 51.61it/s] 
2022-02-25 17:42:26,390 - root - INFO - Concatenating data

2022-02-25 17:42:44,497 - root - INFO - All reviews loaded
```

Reply:
That app has a very broken CSV file for its reviews - lots of unusual characters seem to have bricked it (and it would take some time to repair given size of file). Not urgent given nature of app and behaviour is acceptable